src/drain-service/drain_training_inferencing.py

```diff
@@ -0,0 +1,345 @@
+                    ),
+                ):
+                    action, result = result.popitem()
+                    if not ok:
+                        logging.error("failed to %s document %s" % ())
+            except BulkIndexError as exception:
+                logging.error("Failed to index data")
+                logging.error(exception)
+            logging.info("Updated {} anomalies in ES".format(len(anomaly_df)))
+
+        try:
+            # update normal logs in ES
+            async for ok, result in async_streaming_bulk(
+                es,
+                doc_generator(
+                    df[
+                        [
+                            "_id",
+                            "_op_type",
+                            "_index",
+                            "drain_matched_template_id",
+                            "drain_matched_template_support",
+                        ]
+                    ]
+                ),
+            ):
+                action, result = result.popitem()
+                if not ok:
+                    logging.error("failed to %s document %s" % ())
+        except BulkIndexError as exception:
+            logging.error("Failed to index data")
+            logging.error(exception)
+        logging.info("Updated {} logs in ES".format(len(df)))
+
+
+async def training_signal_check(nw):
+    def weighted_avg_and_std(values, weights):
+        average = np.average(values, weights=weights)
+        # Fast and numerically precise:
+        variance = np.average((values - average) ** 2, weights=weights)
+        return average, math.sqrt(variance)
+
+    iteration = 0
+    train_on_next_chance = True
+    stable = False
+    training_start_ts_ns = time.time_ns()
+    training_end_ts_ns = -1.0
+    very_first_ts_ns = training_start_ts_ns
+    previous_weighted_vol = -1.0
+
+    normal_periods = []
+
+    while True:
+        await asyncio.sleep(20)
+        iteration += 1
+        num_total_clusters_tracking_queue.appendleft(len(template_miner.drain.clusters))
+        logging.info(
+            "training_signal_check: num_total_clusters_tracking_queue {}".format(
+                num_total_clusters_tracking_queue
+            )
+        )
+        num_clusters = np.array(num_total_clusters_tracking_queue)
+        vol = np.std(num_clusters) / np.mean(num_clusters[:10])
+        time_steps = num_clusters.shape[0]
+        weights = np.flip(np.true_divide(np.arange(1, time_steps + 1), time_steps))
+        weighted_mean, weighted_std = weighted_avg_and_std(num_clusters, weights)
+        weighted_vol = weighted_std / np.mean(num_clusters[:10])
+        logging.info(
+            "ITERATION {}: vol= {} weighted_vol= {} normal_periods={}".format(
+                iteration, vol, weighted_vol, str(normal_periods)
+            )
+        )
+
+        if len(num_total_clusters_tracking_queue) > 3:
+            if weighted_vol >= 0.199:
```

Comment:
Fine for now but may be good in the future to give the user an option to set this variable as well if they would like to do so.

Reply:
What specifically? The minimum number of time periods (3 20-second periods right now) elapsed before checking volatility? Or the 0.199 value?
can you name an example in which the user would want to set this value? 
I was thinking it would be better to let them select a low/medium/high sensitivity rather than tweak this value explicitly.